pkg/kapis/devops/v1alpha2/devops.go

```diff
@@ -92,6 +93,12 @@ func (h *ProjectPipelineHandler) StopPipeline(req *restful.Request, resp *restfu
 		return
 	}
 
+	if event := request.AuditEventFrom(req.Request.Context()); event != nil {
```

Comment:
Why auditing has to collect events inside devops module, while all the information needed can be extracted from request?

Reply:
The most auditing information of most APIs can be collected through request and response in auditing component. For a few non-standard APIs, some audit information needs to be obtained through special processing of request and response, such as parsing the resource name from the response. If these special processing are placed in auditing component, then each request will match the request url at least once. Putting special processing of a few non-standard apis into the api handler will reduce the performance impact of the audit on the apiserver.

Reply:
Auditing should happen outside of the business module, decoupled with them, try to think what if we add more modules here(what we will do in v4.0, third party modules), do we need to write auditing logic for each of them inside that module? We will figure out other ways to reduce performance impact, but mixing auditing code with a specific module is not a good path we are going to take.  For non-standard APIs, we can make them become standard step by step.

Reply:
If all apis are standard, auditing will happen outside of the business module, decoupled with them. But there was som few non-standard APIs in 3.0, it is unreasonable to affect the performance of most standard APIs because of a small number of non-standard APIs, 
API should bear the responsibility of providing audit information because it is non-standard. 
This can also promote the standardization process of apis. 
We can optimize or even reconstruct auditing components in 3.1, or when all interfaces are standardized in 4.0, this problem does not exist. This is currently the fastest way to solve this problem.

Reply:
Are there any statistics or experiment data on how much performance impact we have to suffer if we do it the other way?

Reply:
We can test it in 3.1 to see if there are other solutions. The main goal should now be to release version 3.0.

Reply:
I think we should do it right in the first place, the current method is highly coupled with specific modules. @rayzhou2017 how do you think? Should we change it now or leave it to next version?

Reply:
How long will it take if we decouple them, @wanjunlei ? It is NOT good to invade into other business code in terms of architecture. It is like logging. In cloud native application, it throws  logs into console, then logging component is responsible to pick up the logs. The component never cracks into the application to get logs. The same principle should go with auditing in my opinion. It must be decoupling. The only thing is how long it takes that we need to evaluate.

Reply:
/cc @benjaminhuo please also help check this issue

Reply:
Would a sidecar solve the problem?

Reply:
If it depends on the standardization of KS API， then maybe we can just not add audit for non-standard API in KS v3.0. And create an issue to track or discuss with @wansir to standardize KS API.

For now, we can just add audit for standard KS API.

@wanjunlei @zryfish @rayzhou2017

Reply:
OK, I got you. Now I have a suggestion. How serious it will impact in terms of performance for now? If it is not too bad, we can accept the current solution right now since it audits the components which is necessary for users. BUT, we need to create an issue tracking this PR for code refactoring later on. On the other hand, if it affects performance a lot, then do not accept the current solution, and create an issue to track the problem as well. So make a decision.

Reply:
@rayzhou2017 According to @wanjunlei , current solution in this PR have minimum performance impact , but it's indeed a kind of invasion to devops code. The alternative method is to move these non-standard audit to audit module and configure rules for these non-standard components apis and this will impact the performance a bit because audit module has to find out what kind of request it is auditing and then choose corresponding auditing methods.

Reply:
Of course, the better way is to have standard KS APIs for all reqeusts. But most likely this is a long process and we need to discuss with @wansir  or @shaowenchen  together

Reply:
using rules will decouple them, right? Is it the final decoupling solution?

Reply:
This is one of decoupling solution,i will perfect this solution and discuss it later.

Reply:
Now I will delete the code inside devops module and iam module. This pr only handler one situation，
get resource name from request body when request is a resource creating request and the resource is k8s resource or crd. 
The non-standard apis will be handled later.